src/optical_spi.c

```diff
@@ -2,143 +2,146 @@
-        uint32_t timeout = UINT32_MAX;
-        while ((timeout > 0) && (opt_spi_num_bytes < 3)) {
-            timeout--;
-        }
-        print("timeout = %u\n", timeout);
-
-        // If we didn't get 3 bytes
-        if (opt_spi_num_bytes < 3) {
-            //_delay_ms(5000); // long delay = timeout, used for testing
-            return 0;
-        }
 
-        // If we have received all 3 bytes now
-        uint32_t read_data = opt_spi_data;
-        // print ("Received optical data: 0x%06lx\n", read_data);
+// returns status of DATA_RDY pin
+uint8_t get_data_pin(void){
+    return OPT_DATA_PIN >> OPT_DATA;
+}
 
-        // Clear SPI request variables
-        opt_spi_num_bytes = 0;
-        opt_spi_data = 0;
 
-        return read_data;
-    }
+// sends 2 bytes to request reading
+// byte 1 = cmd_byte
+// byte 2 = well_into
+void send_opt_spi_cmd(uint8_t cmd_opcode, uint8_t well_info) {
+    // print("Sending optical the command code %X\n", cmd_opcode);
 
-    // Silence missing return warning
-    return 0;
-}
+    opt_spi_data = 0;
+    spi_frame_number = 0;
 
-/*
-Sends a command to read data for the specified field number from PAY-Optical.
-*/
-void send_opt_spi_cmd(uint8_t field_num) {
-    // print("Sending optical command - field #%u\n", field_num);
+    // Send the command to PAY-Optical to start reading data    
+    set_cs_low(OPT_CS_PIN, &OPT_CS_PORT);
+    send_spi(cmd_opcode);
+    set_cs_high(OPT_CS_PIN, &OPT_CS_PORT);
 
-    opt_spi_data = 0;
-    opt_spi_num_bytes = 0;
+    // wait for DATA_RDYn to go LOW, signaling optical is done reading byte 1
+    // --> aka loop until DATA_RDYn is no longer HIGH
+    uint16_t timeout = UINT16_MAX;
+    while (get_data_pin() == 1 && timeout>0){
+        timeout--;
+    }
 
-    // Send the command to PAY-Optical to start reading data
+    // print("Sending optical the well data %X\n", well_info);
+    // send PAY-OPTICAL the 2nd byte, containing details about the well and type of reading to take
     set_cs_low(OPT_CS_PIN, &OPT_CS_PORT);
-    uint8_t spi_tx = (0b11 << 6) | field_num;
-    send_spi(spi_tx);
+    send_spi(well_info);
     set_cs_high(OPT_CS_PIN, &OPT_CS_PORT);
 
-    // After this, PAY-Optical should set DATA_RDY high, which will trigger the
-    // interrupt below to receive the data
+    // set SPI status to "waiting for OPTICAL to respond"
+    spi_in_progress = true;
 }
 
 
-/*
-This interrupt is triggered when the DATA_RDY pin goes set_cs_high
-(set by PAY-Optical to indicate data is ready).
-The SPI transfer should already be in progress if we get this interrupt.
-*/
-ISR(INT1_vect) {
-    // print("INT1 - DATA_RDY\n");
+// function called by PAY-SSM in its main loop 'every once in a while' to check up on OPTICAL after it sent it a command
+// if PAY-OPTICAL responded properly, proceed with multiple byte SPI transmission
+// otherwise, do nothing
+uint32_t check_received_opt_data(uint8_t num_expected_bytes){
```

Comment:
This function won't return anything

Reply:
I think `num_expected_bytes` would always be 3? Is there a reason you would want it to be variable?

Reply:
voltage